cloud/services/object_storage_objects.go

```diff
@@ -0,0 +1,137 @@
+package services
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	s3manager "github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/smithy-go"
+	"github.com/go-logr/logr"
+
+	"github.com/linode/cluster-api-provider-linode/cloud/scope"
+)
+
+func CreateObject(ctx context.Context, mscope *scope.MachineScope, data []byte, logger logr.Logger) (string, error) {
+	logger.Info("Create Object Storage object")
+
+	if mscope == nil {
+		return "", errors.New("machine scope can't be nil")
+	}
+
+	if mscope.S3Client == nil {
+		return "", errors.New("nil S3 client in machine scope")
+	}
+
+	if mscope.LinodeCluster.Spec.ObjectStore == nil {
+		return "", errors.New("nil cluster object store")
+	}
+
+	if len(data) == 0 {
+		return "", errors.New("got empty data")
+	}
+
+	bucket, err := mscope.GetBucketName(ctx)
+	if err != nil || bucket == "" {
+		return "", errors.New("no bucket name")
+	}
+	// Key by UUID for shared buckets.
+	key := string(mscope.LinodeMachine.ObjectMeta.UID)
+
+	if _, err := mscope.S3Client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+		Body:   s3manager.ReadSeekCloser(bytes.NewReader(data)),
+		// TODO: Enable SSE?
+		// ServerSideEncryption: types.ServerSideEncryptionAes256,
```

Comment:
I would enable it unless we have to pass an encryption key for this, I forget if that's necessary since I looked at using SSE a bit over a year ago.

Reply:
It looks like for the Linode Object Storage provided method ([SSE-C](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html)), the client needs to set specific SSE-related headers in the request ([1](https://aws.amazon.com/blogs/developer/generating-amazon-s3-pre-signed-urls-with-sse-c-part-4/), [2](https://www.linode.com/community/questions/22648/download-from-object-storage-using-sse-c-signed-urls)). This level of granularity is currently possible in cloud-init's [include file](https://cloudinit.readthedocs.io/en/latest/explanation/format.html#include-file) user data format.